### PR TITLE
Rework and simplify selection in TermControl

### LINKED ISSUE #4082: How to drag a file into a window?
I want to use " ADB install " to install the APK file. When I drag the file to the window, a reject icon will be displayed...

### LINKED ISSUE #4477: [ui] Difficult to get rid of selection area
In Ubuntu Terminal selection caret is not the default mode. In Windows Terminal it is hard to get rid of selection mode. Even mouse clicking away on empty area preserves it (right mouse click or Esc does the trick, but we don't need these tricks in other applications or terminals). In many terminals a single clicking is not enough to turn on selection and snapping is needed.

image

Strangely, hitting Ctrl+Shift+C removes selection area after copying to clipboard. This is quite annoying, since often we want to preserve the selected area (for copying again later or checking what exactly was being copied)

(Also having a typing caret and selection area active at the same time is strange, e.g. in Chrome in text fields it is not like that, though this is less critical)

Original umbrella issue: https://github.com/microsoft/terminal/issues/2209

### LINKED ISSUE #5082: Determine proper behavior for single-cell selection; only highlight on click + threshold-drag?
# Environment

```
Windows 10 v1909
Windows Terminal 0.10.781.0
```

# Steps to reproduce

Click anywhere in the text on an already focused Windows Terminal window.

# Expected behavior

Window should do nothing since it is already focused. Text should not be highlighted because there has been no mouse drag.

# Actual behavior

The character that's clicked on is now going to overwrite the clipboard if I right click to paste something I've just hunted out in another program or window. I will often fail to notice I've highlighted anything, and even if I do catch myself before muscle memory has completed the action, I cannot "unhighlight" text, and am basically screwed at this point. This means I have to go through with the failed paste operation, if I haven't already, and then go hunt out what I was about to paste again.

This is a pet hate of mine. Please make it so that clicking on the window is not interpreted as marking text, or at least provide an option for this. You would simply have to wait for a mouse drag event, which is what people generally do when they DO want to copy some text, so most people would not even notice this change unless they often copy a single character from a console.

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
This commit rewrites selection handling at the TermControl layer.
Previously, we were keeping track of a number of redundant variables
that were easy to get out of sync.

The new selection model is as follows:

* A single left click will always begin a _pending_ selection operation
* A single left click will always clear a selection (#4477)
* A double left click will always begin a word selection
* A triple left click will always begin a line selection
* A selection will only truly start when the cursor moves a quarter of
  the smallest dimension of a cell (usually its width) in any direction
  _This eliminates the selection of a single cell on one click._
  (#4282, #5082)
* We now keep track of whether the selection has been "copied", or
  "updated" since it was last copied. If an endpoint moves, it is
  updated. For copy-on-select, it is only copied if it's updated.
  (#4740)

Because of this, we can stop tracking the position of the focus-raising
click, and whether it was part of click-drag operation. All clicks can
_become_ part of a click-drag operation if the user drags.

We can also eliminate the special handling of single cell selection at
the TerminalCore layer: since TermControl determines when to begin a
selection, TerminalCore no longer needs to know whether copy on select
is enabled _or_ whether the user has started and then backtracked over a
single cell. This is now implicit in TermControl.

Fixes #5082; Fixes #4477

### CODE REVIEW

**Comment by carlos-zamora** [comment]
> * A single left click will always clear a selection (#4477)

This one is a bit concerning. Really only because ConHost does _not_ do this. I hate to say it, but it makes me wonder if we should make this particular setting a pointer binding in #1553

> * A selection will only truly start when the cursor moves a quarter of
>   the smallest dimension of a cell (usually its width) in any direction
>   _This eliminates the selection of a single cell on one click._
>   (#4282, #5082)
> * We now keep track of whether the selection has been "copied", or
>   "updated" since it was last copied. If an endpoint moves, it is
>   updated. For copy-on-select, it is only copied if it's updated.
>   (#4740)
>  * [Paraphrasing:] we can stop tracking the position of the focus-raising
> click, and whether it was part of click-drag operation.
> We can also eliminate the special handling of single cell selection at
> the TerminalCore layer.

These are all things I really like and want. I think the only reason we needed `copyOnSelect` in TermCore was for the single cell selection. But if we can differentiate between a "tap" pending a selection action vs a "drag" actually selecting, then this model looks a lot cleaner.


Side note: #4082 is probably not the issue you want to link here?

**Comment by DHowett-MSFT** [comment]
Yeah, I meant #5082. Let’s talk about selection clearing: why do we want to keep the selection when you click somewhere else? All other text buffer applications clear it to start another selection. Right now there’s no way to even abort a selection with the mouse (see complaints in 4477)

**Comment by DHowett-MSFT** [comment]
I legit think “because conhost did it” is an okay reason for _some_ things, but Terminal also presents a chance to change the things conhost did that don’t make sense. Right now, conhost always starts and completes a selection on a single click so there isn’t a clear state—_all_ you can do is move to another single cell selection, which we’ve established isn’t useful :smile:

If people are using single cell selection to pause output, that’s a failing on our part for us having not provided any other way to pause.

**Comment by carlos-zamora** [comment]
> I legit think “because conhost did it” is an okay reason for _some_ things, but Terminal also presents a chance to change the things conhost did that don’t make sense. Right now, conhost always starts and completes a selection on a single click so there isn’t a clear state—_all_ you can do is move to another single cell selection, which we’ve established isn’t useful 😄
> 
> If people are using single cell selection to pause output, that’s a failing on our part for us having not provided any other way to pause.

I agree and I think this updated behavior is more correct. Rather, my only concern is that we'll have a number of users coming from ConHost expecting it to work somewhat similarly. That said, we could do what we've done with other features and introduce this new behavior as the default. Then if somebody submits an issue expecting it to work the other way, we _then_ address it by adding it in as a customization option.

I genuinely think what you are proposing is the right behavior. I just don't want people familiar with Windows-isms to suddenly get confused. Even though I think this is one of those bad Windows-isms.

**Comment by DHowett-MSFT** [comment]
Also, this diff is +35 -120 :smile: you know how much I love deleting code

File: src/cascadia/TerminalControl/TermControl.h [hunk 1]
```diff
@@ -155,12 +155,11 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
         // Used for PointerPoint.Timestamp Property (https://docs.microsoft.com/en-us/uwp/api/windows.ui.input.pointerpoint.timestamp#Windows_UI_Input_PointerPoint_Timestamp)
         Timestamp _multiClickTimer;
-        Timestamp _lastMouseClickTimestamp;
         unsigned int _multiClickCounter;
+        Timestamp _lastMouseClickTimestamp;
         std::optional<winrt::Windows::Foundation::Point> _lastMouseClickPos;
-
-        std::optional<winrt::Windows::Foundation::Point> _focusRaisedClickPos;
-        bool _clickDrag;
+        std::optional<winrt::Windows::Foundation::Point> _singleClickTouchdownPos;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.h]: 🏈📣

(I had to)

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -1028,16 +1023,22 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
-                // from out of focus, so PointerMoved has to set it.
-                if (_focusRaisedClickPos)
+                if (_singleClickTouchdownPos)
                 {
-                    _terminal->SetSelectionAnchor(_GetTerminalPosition(*_focusRaisedClickPos));
+                    // Figure out if the user's moved a quarter of a cell's smaller axis away from the clickdown point
+                    auto& touchdownPoint{ *_singleClickTouchdownPos };
+                    auto distance{ std::sqrtf(std::powf(cursorPosition.X - touchdownPoint.X, 2) + std::powf(cursorPosition.Y - touchdownPoint.Y, 2)) };
+                    const auto fontSize{ _actualFont.GetSize() };
+                    if (distance >= (std::min(fontSize.X, fontSize.Y) / 4.f))
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.cpp]: could/should we cache this? Also, isn't X always smaller than Y?

File: src/cascadia/TerminalControl/TermControl.h [hunk 3]
```diff
@@ -155,12 +155,11 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+        bool _selectionUpdatedThisCycle;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.h]: Could you add a description here or somewhere for this new bool?

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 2 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/TermControl.cpp]: X isn't always smaller than Y; some fonts are strange. Consider that caching requires cache invalidation, which is a lot to ask for a few simple floating point operations.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 4]
```diff
@@ -980,7 +975,7 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
             else if (point.Properties().IsRightButtonPressed())
             {
                 // CopyOnSelect right click always pastes
-                if (_terminal->IsCopyOnSelectActive() || !_terminal->IsSelectionActive())
+                if (!_selectionUpdatedThisCycle || !_terminal->IsSelectionActive())
```
Comment (leonMSFT) [inline src/cascadia/TerminalControl/TermControl.cpp]: I'm having a bit of trouble understanding why the change in this if statement is still equivalent to the comment above of "CopyOnSelect right click always pastes", can you explain?

Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/TermControl.cpp]: Iiiii might have forgotten to push the commit that changed the condition here back to CoS 😅

**Comment by DHowett-MSFT** [comment]
@leonMSFT after playing with it, any strong feelings on whether focusing a window with a selection should preserve or destroy the selection? Notepad/word destroy it; xterm destroys it; putty destroys it

**Comment by leonMSFT** [comment]
I think it's fine to clear selection when focusing, makes the click-to-clear selection behavior more consistent 👍

**Comment by aybiss** [comment]
@DHowett-MSFT hey so I'm stoked you implemented my idea so thanks!

When will this make its way into a release? Do I just wait for the app to update from the store or what? Sorry I've literally used the MS store only once before this! :-)

**Comment by DHowett-MSFT** [comment]
@aybiss thanks! So, we're trying to release _approximately_ monthly, and since this is a behavior change we're probably going to hold it until the next preview build when we change the version number. It should not be _too_ long :smile:
